Mathlib/Analysis/SpecialFunctions/BinEntropy.lean

```diff
@@ -0,0 +1,748 @@
+variable {F : Type*} [NormedAddCommGroup F] [NormedSpace 𝕜 F]
+variable {f f₀ f₁ g : E → F}
+variable {x : E}
+
+-- TODO clean up, put somewhere else?
+lemma not_DifferentiableAt_of_not_DifferentiableAt_add_DifferentiableAt_
+    (hf : ¬ DifferentiableAt 𝕜 f x) (hg : DifferentiableAt 𝕜 g x) :
+    ¬ DifferentiableAt 𝕜 (fun y => f y + g y) x := by
+  have f_eq_sum_sub_g: f = (fun y => f y + g y) - g := by
+    ext
+    simp only [Pi.sub_apply, add_sub_cancel_right]
+  by_contra
+  have : DifferentiableAt 𝕜 f x := by
+    rw [f_eq_sum_sub_g]
+    apply DifferentiableAt.sub
+    · simp
+      assumption
+    · assumption
+  contradiction
+
+-- TODO clean up, put somewhere else?
+lemma not_DifferentiableAt_of_DifferentiableAt_add_not_DifferentiableAt
+    (hf : DifferentiableAt 𝕜 f x) (hg : ¬ DifferentiableAt 𝕜 g x) :
+     ¬ DifferentiableAt 𝕜 (fun y => f y + g y) x := by
+  rw [show (fun y ↦ f y + g y) = (fun y ↦ g y + f y) by ext; rw [add_comm]]
+  exact not_DifferentiableAt_of_not_DifferentiableAt_add_DifferentiableAt_ hg hf
+
+lemma differentiableAt_iff_differentiableAt_comp_mul_add
+    {a b m : 𝕜} (hm : m ≠ 0) (f : 𝕜 → E) :
+    DifferentiableAt 𝕜 f a ↔ DifferentiableAt 𝕜 (fun x => f (m * x + b)) (m⁻¹ * (a - b)):= by
+  constructor <;> intro h
+  · apply DifferentiableAt.comp
+    have : (m * (m⁻¹ * (a - b)) + b) = a := by
+      simp_all only [ne_eq, isUnit_iff_ne_zero, not_false_eq_true, IsUnit.mul_inv_cancel_left,
+        sub_add_cancel]
+    rw [this]
+    exact h
+    apply DifferentiableAt.add
+    · exact DifferentiableAt.mul (differentiableAt_const m) differentiableAt_id'
+    · exact differentiableAt_const b
+  · have diff_affine : DifferentiableAt 𝕜 (fun x => m⁻¹ * (x - b)) a := by
+      apply DifferentiableAt.mul (differentiableAt_const m⁻¹)
+      apply DifferentiableAt.sub_const differentiableAt_id'
+    have : f = (fun x ↦ f (m * x + b)) ∘ (fun x => m⁻¹ * (x - b)) := by
+        ext
+        simp only [Function.comp_apply]
+        field_simp
+    rw [this]
+    apply DifferentiableAt.comp
+    exact h
+    exact diff_affine
+
+@[simp]
+lemma fderiv_deriv' {f : 𝕜 → 𝕜} {x y : 𝕜} : (fderiv 𝕜 f x : 𝕜 → 𝕜) y = (deriv f x) * y := by
+  rw [← deriv_fderiv]
+  simp only [ContinuousLinearMap.smulRight_apply, ContinuousLinearMap.one_apply, smul_eq_mul]
+  ring
+
+end general
+
+lemma differentiableAt_binaryEntropy {x : ℝ} (xne0: x ≠ 0) (gne1 : x ≠ 1) :
+    DifferentiableAt ℝ (fun p => -p * log p - (1 - p) * log (1 - p)) x := by
+  apply DifferentiableAt.sub
+  apply DifferentiableAt.mul
+  apply DifferentiableAt.neg
+  exact differentiableAt_id'
+  apply DifferentiableAt.log differentiableAt_id' xne0
+  apply DifferentiableAt.mul
+  apply DifferentiableAt.sub
+  apply differentiableAt_const
+  exact differentiableAt_id'
+  apply DifferentiableAt.log
+  apply DifferentiableAt.sub
+  apply differentiableAt_const
+  exact differentiableAt_id'
+  exact sub_ne_zero.mpr gne1.symm
```

Comment:
```suggestion
  refine DifferentiableAt.sub (DifferentiableAt.mul (by fun_prop) ?_)
      (DifferentiableAt.mul (by fun_prop)
        (DifferentiableAt.log (by fun_prop) (sub_ne_zero.mpr gne1.symm)))
  exact DifferentiableAt.log differentiableAt_id' xne0
```
It turns out that `fun_prop` can handle some of the smaller subgoals. Note that the original proof would have needed bullet points to separate subgoals.